### PR TITLE
Bugfix FXIOS-9289 [Multi-window] Fix blur when backgrounding app

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9289)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20569)

## :bulb: Description

In practice iPadOS does not update the window thumbnails in time for the blur to be visible in the app switcher. The fix here is to also blur when the app resigns active status, and to remove the blur (if authenticated) when active status is regained. If the loss of active status is also coupled with a change in foreground/background status then the normal authentication logic will run as usual.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by dataports** [comment]
Not sure if this is an edge case due to using simulator (I don't have an iPad setup to test on currently) but I found when testing on simulator that if I put the application to the app switcher and brought it back shortly after, the blur would still be there. If I set breakpoints in the observers in `viewWillAppear` the issue would stop, so maybe it's something to do with timing?

https://github.com/mozilla-mobile/firefox-ios/assets/5545720/9d7542d0-fd5a-49ef-8a06-2a18dff31bac

File: firefox-ios/Client/Frontend/AuthenticationManager/SensitiveViewController.swift [hunk 1]
```diff
@@ -8,60 +8,71 @@ import Shared
...
-        ) { [weak self] notification in
+        didEnterBackgroundObserver = observe(UIApplication.didEnterBackgroundNotification) { [weak self] notification in
             guard let self = self else { return }
-
             self.isAuthenticated = false
             self.installBlurredOverlay()
         }
+
+        willResignActiveObserver = observe(UIApplication.willResignActiveNotification) { [weak self] notification in
+            guard let self = self else { return }
```
Comment (dataports) [inline firefox-ios/Client/Frontend/AuthenticationManager/SensitiveViewController.swift:34]: Could use `guard let self else { return }` in this PR if you want, not sure if we prefer using one or the other in the codebase, there seems to be instances of both

**Comment by mattreaganmozilla** [comment]
@dataports Thanks, I'm taking a look, should have an update in a bit

**Comment by mattreaganmozilla** [comment]
Update: I'm still investigating this. It's a bit more complicated than it appears; the only notification hook I've found so far that addresses the original issue (which appears to be an iPadOS bug) is willResignActive. The difficulty is that it's not really the notification we want to use and is called in a variety of other scenarios also (such as when the device passcode overlay is presented atop the list). And it ends up conflicting with the other notifications, which are still necessary since those are the principal hooks we use to determine when to (de)authenticate. Will post another update as soon as I can dig into this a bit further.

**Comment by mattreaganmozilla** [comment]
@dataports Can you take another look? I've pushed up a fix that should address the previous conflicts/issues.

File: firefox-ios/Client/Frontend/AuthenticationManager/SensitiveViewController.swift [hunk 2]
```diff
@@ -6,62 +6,86 @@ import UIKit
...
+        guard !isAuthenticated else { return }
+        AppAuthenticator().authenticateWithDeviceOwnerAuthentication { [self] result in
+            switch result {
+            case .success:
+                isAuthenticated = false
+                removedBlurredOverlay()
+            case .failure:
+                isAuthenticated = false
+                navigationController?.dismiss(animated: true, completion: nil)
+                dismiss(animated: true)
```
Comment (mattreaganmozilla) [inline firefox-ios/Client/Frontend/AuthenticationManager/SensitiveViewController.swift:88]: Note: no changes here, only refactored to change `if` to `guard`

**Comment by mattreaganmozilla** [comment]
@Mergifyio backport release/v128

**Review by dataports — APPROVED** [review]
Looks great!

**Comment by mattreaganmozilla** [comment]
@dataports Thanks very much for testing out the branch
